Stabilize cfg(doc) (#61351)
rust-lang/rust: src/doc/rustdoc/src/the-doc-attribute.md

Since primitive types are defined in the compiler, there's no place to attach documentation
attributes. This attribute is used by the standard library to provide a way to generate
documentation for primitive types.

> ## `#[cfg(rustdoc)]`: Documenting platform-/feature-specific information

Reviewer: I just realized something about these docs. First, the heading here is a little misleading, since `cfg(rustdoc)` isn't just about platform-specific stuff. It's feasible that people could "invent" items that only exist for documentation, for whatever reason. However, writing something like "exposing items to documentation" feels *more* misleading, since it implies that you need that to add items to docs in the first place. I think this header text is acceptable, since platform-specific docs was the driving force for it, but it's something to consider.

Second, there isn't really a good chapter for these docs. This chapter is specifically about the `#[doc]` attribute, but this section is about something you add to the `#[cfg]` attribute. However, all the other chapters make even less sense to add this to. I'm wondering whether we should create a new chapter that talks about "advanced" features that are stable. I'm not sure what else would go with it, though... :thinking: Maybe we can just leave this here. I'm curious what @ollie27 and @steveklabnik think, though.

> Because of the way Rustdoc documents a crate, the documentation it creates is specific to the target

Reviewer: How about this for rewriting these docs:

> For conditional compilation, Rustdoc treats your crate the same way the compiler does: Only things from the host target are available (or from the given `--target` if present), and everything else is "filtered out" from the crate. This can cause problems if your crate is providing different things on different targets, and you want your documentation to reflect all the available items you provide.
>
> If you want to make sure an item is seen by Rustdoc regardless of what platform it's being run from, you can apply `#[cfg(rustdoc)]` to it. Rustdoc sets this whenever it's building documentation, so anything that uses that flag will make it into documentation it generates. To apply this to an item with other `#[cfg]` filters on it, you can write something like `#[cfg(any(windows, rustdoc))]`. This will preserve the item either when built normally on Windows, or when being documented anywhere. For example:
>
> (current code sample)
>
> Here, the respective tokens can only be used by dependent crates on their respective platforms, but they will both appear in documentation.

Author: I really prefer your version so I'll just use it as is if you don't mind.

rustc compiles for. Anything that's specific to any other target is dropped via `#[cfg]` attribute
processing early in the compilation process. However, Rustdoc has a trick up its sleeve to handle
platform-specific code if it *does* receive it.

For Rustdoc to document an item, it needs to see it, regardless of what platform it's currently
running on. To aid this, Rustdoc sets the flag `#[cfg(rustdoc)]` when running on your crate.
Combining this with the target platform of a given item allows it to appear when building your crate
normally on that platform, as well as when building documentation anywhere.

For example, `#[cfg(any(windows, rustdoc))]` will preserve the item either on Windows or during the
documentation process. For example:

```rust
/// Token struct that can only be used on Windows.

> #[cfg(any(windows, all(windows, rustdoc)))]

Reviewer: These samples should just use `#[cfg(any(platform, rustdoc))]`, to match the text above it. As written, they will still only appear when generating the docs using a `--target` for their selected platform. (The current cfg attributes are like, "this should only appear *either* on Windows/Unix, *or* when *both* Windows/Unix and `rustdoc` are set".)

Author: Agreed, that even simplifies the example. Updating it.

Since primitive types are defined in the compiler, there's no place to attach documentation
attributes. This attribute is used by the standard library to provide a way to generate
documentation for primitive types.

## `#[cfg(rustdoc)]`: Documenting platform-/feature-specific information

For conditional compilation, Rustdoc treats your crate the same way the compiler does: Only things
from the host target are available (or from the given `--target` if present), and everything else is
"filtered out" from the crate. This can cause problems if your crate is providing different things
on different targets and you want your documentation to reflect all the available items you
provide.

> If you want to make sure an item is seen by Rustdoc regardless of what platform it's being run from,

Reviewer 2: ```suggestion
If you want to make sure an item is seen by Rustdoc regardless of what platform it's targeting,
```

Since primitive types are defined in the compiler, there's no place to attach documentation
attributes. This attribute is used by the standard library to provide a way to generate
documentation for primitive types.

## `#[doc(cfg(...))]` / `#[cfg(rustdoc)]`: Documenting platform-/feature-specific information

Because of the way Rustdoc documents a crate, the documentation it creates is specific to the target
rustc compiles for. Anything that's specific to any other target is dropped via `#[cfg]` attribute
processing early in the compilation process. However, Rustdoc has a trick up its sleeve to handle
platform-specific code if it *does* receive it.

> Because Rustdoc doesn't need to fully compile a crate to binary, it replaces function bodies with

Reviewer 3: Unsure if this implementation detail needs to be exposed.

Author: I found it interesting so I didn't remove it. :)

> `loop {}` to prevent having to process more than necessary. This means that any code within a

Reviewer 4: How does this cope with functions that have `-> impl Trait` where `!` does not implement `Trait`?

Author: cc @QuietMisdreavus

Reviewer 5: There is currently an exception for `-> impl Trait`: https://github.com/rust-lang/rust/blob/26451ef7b5e00887dc8f27717ff34262df23d655/src/librustc_interface/util.rs#L668-L673

This means that `#[doc(cfg(...))]` doesn't work for those functions yet so I don't think this feature is ready for stabilization yet.

Reviewer 4: ```rust
//  * `impl Trait`, due to issue #43869 that functions returning impl Trait cannot be diverging.
//    Solving this may require `!` to implement every trait, which relies on the an even more
//    ambitious form of the closed RFC #1637. See also [#34511].
```
This bit seems unlikely from my POV as a T-Lang member... cc @rust-lang/lang if y'all have thoughts...

Reviewer: I'm with @ollie27 here, i had assumed the reason we couldn't stabilize `doc(cfg)` yet was because the implementation didn't work for all situations.
